File: backend/app/schemas/job.py
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel, Field


class JobPostOut(BaseModel):
    post_name: str
    vacancies: int = 0
    pay_level: str | None = None


class JobDateOut(BaseModel):
    event_key: str
    event_date: date


class JobOut(BaseModel):
    id: str
    slug: str
    title: str
    dept: str | None = None
    category: str | None = None
    state_codes: list[str] = Field(default_factory=list)
    vacancies: int | None = None
    qualification: str | None = None
    salary: str | None = None
    age_limit: str | None = None
    last_date: date | None = None
    apply_url: str | None = None
    pdf_url: str | None = None
    status: str = "live"
    is_sponsored: bool = False
    published_at: datetime | None = None
    verified_at: datetime | None = None
    updated_at: datetime | None = None
    document_type: str | None = None
    verification_status: str | None = None
    completeness_score: int | None = None
    publication_confidence: float | None = None
    published_to_site: bool | None = None
    primary_pdf_url: str | None = None
    post_name: str | None = None
    posts: list[JobPostOut] = Field(default_factory=list)
    important_dates: list[JobDateOut] = Field(default_factory=list)
    detail: dict[str, Any] = Field(default_factory=dict)


class JobListResponse(BaseModel):
    items: list[JobOut]
    total: int
    limit: int
    offset: int
    page: int = 1
    total_pages: int = 1
    degraded: bool = False

File: frontend/src/types/job.ts
/** Shared job shapes used across cards, detail, and adapters. */

export type JobContentSection = {
  heading?: string
  paragraphs?: string[]
  lists?: string[][]
  tables?: Record<string, string>[][]
  links?: { label?: string; url?: string }[]
}

export type JobPostRow = {
  post_name?: string
  post?: string
  vacancies?: number
  pay_level?: string
}

export type JobDateRow = {
  event_key?: string
  event?: string
  event_date?: string
  date?: string
}

export type JobDetailBlob = {
  pdfUrls?: string[]
  pdf_urls?: string[]
  pdfUrl?: string
  pdf_url?: string
  summary?: string
  source?: string
  age?: string
  post_name?: string
  posts?: JobPostRow[]
  content_sections?: JobContentSection[]
  streetAddress?: string
  street_address?: string
  postalCode?: string
  postal_code?: string
  pincode?: string
}

export type JobRecord = {
  id?: string
  slug?: string
  title?: string
  post_name?: string
  posts?: JobPostRow[]
  important_dates?: JobDateRow[]
  about?: string
  dept?: string
  category?: string
  state?: string
  stateIds?: string[]
  vacancies?: number
  rawVacancies?: number
  qual?: string
  salary?: string
  age?: string
  selection?: string[]
  howApply?: string[]
  lastDate?: string
  last_date?: string
  publishedDate?: string
  published_at?: string
  updatedDate?: string
  updated_at?: string
  apply_url?: string
  pdfUrl?: string
  pdfUrls?: string[]
  status?: string
  detail?: JobDetailBlob
  _enriched?: boolean
  _fromLive?: boolean
  [key: string]: unknown
}
